File: server/src/net/protocol/serials.py
from ...ack import Ack
from ...bet import Bet
from ...query import Query
from ...response import Response
from .protocol import (
    BYTE_ORDER,
    LEN_BATCH_SIZE,
    LEN_STR_SIZE,
    LEN_WINNER_AMOUNT,
    MsgType,
    LEN_TYPE,
)


class SerialError(Exception):
    pass


def serialize(msg) -> bytes:
    if isinstance(msg, Ack):
        return __serialize_ack(msg)
    elif isinstance(msg, Response):
        return __serialize_response(msg)
    else:
        raise SerialError(f"unsupported serialization for message {msg}")


def deserialize(serial: bytes):
    msg_type = serial[:LEN_TYPE]
    if msg_type == MsgType.TYPE_BET_BATCH.value:
        return __deserialize_bet_batch(serial[LEN_TYPE:])
    elif msg_type == MsgType.TYPE_QUERY.value:
        return __deserialize_query(serial[LEN_TYPE:])
    else:
        raise SerialError(f"unknown message type {msg_type}")


def __serialize_ack(ack: Ack) -> bytes:
    if ack.ok:
        serial = MsgType.TYPE_ACK.value
    else:
        serial = MsgType.TYPE_NACK.value

    return serial


def __serialize_response(response: Response) -> bytes:
    serial = MsgType.TYPE_RESPONSE.value
    serial_winner_amount = response.winner_amount.to_bytes(
        LEN_WINNER_AMOUNT, byteorder=BYTE_ORDER
    )
    serial += serial_winner_amount

    return serial


def __deserialize_bet_batch(serial: bytes) -> list[Bet]:
    ptr = 0
    size = int.from_bytes(serial[ptr:LEN_BATCH_SIZE], byteorder=BYTE_ORDER)
    ptr += LEN_BATCH_SIZE
    bets = []
    for _ in range(0, size):
        bet, consumed = __deserialize_bet(serial[ptr:])
        bets.append(bet)
        ptr += consumed

    return bets


def __deserialize_query(_: bytes) -> Query:
    return Query()


def __deserialize_bet(serial: bytes) -> tuple[Bet, int]:
    """
    Deserializes a byte slice into a bet.

    Returns the deserialized bet along with the bytes it consumed.
    """
    ptr = 0
    # agency, consumed = __deserialize_string(serial[ptr:])
    # ptr += consumed
    first_name, consumed = __deserialize_string(serial[ptr:])
    ptr += consumed
    last_name, consumed = __deserialize_string(serial[ptr:])
    ptr += consumed
    document, consumed = __deserialize_string(serial[ptr:])
    ptr += consumed
    birthdate, consumed = __deserialize_string(serial[ptr:])
    ptr += consumed
    number, consumed = __deserialize_string(serial[ptr:])
    ptr += consumed

    bet = Bet("1", first_name, last_name, document, birthdate, number)  # TODO
    return bet, ptr


def __deserialize_string(serial: bytes) -> tuple[str, int]:
    """
    Deserializes a byte slice into a string.

    Returns the deserialized string along with the bytes it consumed,
    including the ones for its length.
    """
    length = int.from_bytes(serial[:LEN_STR_SIZE], byteorder=BYTE_ORDER)
    len_total = LEN_STR_SIZE + length
    string = serial[LEN_STR_SIZE:len_total].decode()

    return string, len_total

File: client/src/comms/messages/serials.go
package messages

import (
	"bytes"
	"encoding/binary"
	"errors"

	"github.com/7574-sistemas-distribuidos/docker-compose-init/client/src/comms/protocol"
)

func (bets BetBatch) Serialize() []byte {
	var serial bytes.Buffer
	serial.WriteByte(protocol.TYPE_BET_BATCH)
	size := uint16(len(bets))
	binary.Write(&serial, binary.BigEndian, size)
	for _, b := range bets {
		b.serializeInto(&serial)
	}

	return serial.Bytes()
}

func (f Fin) Serialize() []byte {
	serial := []byte{protocol.TYPE_FIN}
	return serial
}

func (q Query) Serialize() []byte {
	serial := []byte{protocol.TYPE_QUERY}
	return serial
}

func Deserialize(bytes []byte) (Message, error) {
	switch bytes[0] {
	case protocol.TYPE_ACK:
		return deserializeAck(bytes[1:])
	case protocol.TYPE_NACK:
		return deserializeNack(bytes[1:])
	case protocol.TYPE_RESPONSE:
		return deserializeResponse(bytes[1:]), nil
	default:
		return Ack{}, errors.New("unknown response type")
	}
}

func deserializeAck(_ []byte) (Ack, error) {
	return Ack{true}, nil
}

func deserializeNack(_ []byte) (Ack, error) {
	return Ack{false}, nil
}

func deserializeResponse(serial []byte) Response {
	ptr := 0
	WinnerAmount := int(binary.BigEndian.Uint16(serial[ptr:protocol.LEN_WINNER_AMOUNT]))
	ptr += protocol.LEN_WINNER_AMOUNT

	Winners := make([]string, 0, WinnerAmount)
	for i := 0; i < WinnerAmount; i++ {
		winner, consumed := deserializeString(serial[ptr:])
		Winners = append(Winners, winner)
		ptr += consumed
	}

	response := Response{Winners}
	return response
}

func (b *Bet) serializeInto(buf *bytes.Buffer) []byte {
	serializeStringInto(b.Agency, buf)
	serializeStringInto(b.FirstName, buf)
	serializeStringInto(b.LastName, buf)
	serializeStringInto(b.Document, buf)
	serializeStringInto(b.BirthDate, buf)
	serializeStringInto(b.Number, buf)

	return buf.Bytes()
}

// Serialize string into bytes and write them into the buffer,
// prepending one byte with the length of the string bytes
func serializeStringInto(s string, buf *bytes.Buffer) {
	serial_len := byte(uint8(len(s)))
	buf.WriteByte(serial_len)
	buf.WriteString(s)
}

// Deserializes a byte slice into a winner document.
//
//	Returns the deserialized document along with the bytes it consumed.
func deserializeString(serial []byte) (string, int) {
	// FIXME: I was not able to do serial[:LEN_STR_SIZE], so this becomes
	// a bug if that constant changes.
	len := int(serial[0])
	len_total := protocol.LEN_STR_SIZE + len
	str := string(serial[protocol.LEN_STR_SIZE:len_total])

	return str, len_total
}
